### PR TITLE
Allow foreign exceptions to unwind through Rust code and Rust panics to unwind through FFI

### LINKED ISSUE #65441: catch_unwind incorrectly tries to stop ForceUnwinds
MWE: 

```rust
use std::os::unix::thread::JoinHandleExt;
use libc::pthread_cancel;

struct DropGuard;
impl Drop for DropGuard {
    fn drop(&mut self) {
        println!("unwinding foo");
    }
}

fn foo() {
    let _x = DropGuard;
    println!("thread started");
    std::thread::sleep(std::time::Duration::new(1, 0));
    println!("thread finished");
}

fn main() {
    let handle0 = std::thread::spawn(foo);
    std::thread::sleep(std::time::Duration::new(0, 10_000));
    unsafe { 
        let x = pthread_cancel(handle0.as_pthread_t());
        assert_eq!(x, 0);
    }
    handle0.join();
}
```

([Playground](https://play.rust-lang.org/?version=nightly&mode=release&edition=2018&gist=bfed948efb5de95c7f35317aa348533c))

Output:

```
thread started
unwinding foo

```

Errors:

```
   Compiling playground v0.0.1 (/playground)
warning: unused `std::result::Result` that must be used
  --> src/main.rs:25:5
   |
25 |     handle0.join();
   |     ^^^^^^^^^^^^^^^
   |
   = note: `#[warn(unused_must_use)]` on by default
   = note: this `Result` may be an `Err` variant, which should be handled

    Finished release [optimized] target(s) in 0.86s
     Running `target/release/playground`
FATAL: exception not rethrown
timeout: the monitored command dumped core
/root/entrypoint.sh: line 8:     7 Aborted                 timeout --signal=KILL ${timeout} "$@"

```

According to the [Itanium ABI](https://itanium-cxx-abi.github.io/cxx-abi/abi-eh.html#base-framework):

> * "forced" unwinding (such as caused by longjmp or thread termination).
> [...]
> During "forced unwinding", on the other hand, an external agent is driving the unwinding. For instance, this can be the longjmp routine. This external agent, not each personality routine, knows when to stop unwinding. The fact that a personality routine is not given a choice about whether unwinding will proceed is indicated by the _UA_FORCE_UNWIND flag.

So `catch_unwind` definitely needs to always catch Rust exceptions, and at the thread boundary it probably also want to catch foreign exceptions and abort since whatever that should do probably won't work. When not on a thread boundary, aborting on foreign exceptions might be a conservative step until a better solution is discussed (an alternative solution would be not to catch them and just let them escape, or to catch them and have the `Any` be some `ForeignPanic` type defined somewhere in `std::panic` that users can use to detect that a foreign exception was raised). What it probably shouldn't do is catch "ForceUnwind" otherwise `longjmp` might not work across it and threads will not be canceled, but... maybe the conservative thing to do would be for this to abort for the time being?

### PR DESCRIPTION
**Pull request by Amanieu**
This PR fixes interactions between Rust panics and foreign (mainly C++) exceptions.

C++ exceptions (and other FFI exceptions) can now safely unwind through Rust code:
- The FFI function causing the unwind must be marked with `#[unwind(allowed)]`. If this is not the case then LLVM may optimize landing pads away with the assumption that they are unreachable.
- Drop code will be executed as the exception unwinds through the stack, as with a Rust panic.
- `catch_unwind` will *not* catch the exception, instead the exception will silently continue unwinding past it.

Rust panics can now safely unwind through C++ code:
- C++ destructors will be called as the stack unwinds.
- The Rust panic can only be caught with `catch (...)`, after which it can be either rethrown or discarded.
- C++ cannot name the type of the Rust exception object used for unwinding, which means that it can't be caught explicitly or have its contents inspected.

Tests have been added to ensure all of the above works correctly.

Some notes about non-C++ exceptions:
- `pthread_cancel` and `pthread_exit` use unwinding on glibc. This has the same behavior as a C++ exception: destructors are run but it cannot be caught by `catch_unwind`.
- `longjmp` on Windows is implemented using unwinding. Destructors are run on MSVC, but not on MinGW. In both cases the unwind cannot be caught by `catch_unwind`.
- As with C++ exceptions, you need to mark the relevant FFI functions with `#[unwind(allowed)]`, otherwise LLVM will optimize out the destructors since they seem unreachable.

I haven't updated any of the documentation, so officially unwinding through FFI is still UB. However this is a step towards making it well-defined.

Fixes #65441

cc @gnzlbg
r? @alexcrichton

### CODE REVIEW

**Comment by Amanieu** [comment]
Also note that I haven't tested this with SEH yet, so the test will probably fail on windows. Emscripten seems to be using some hack for unwinding that I don't fully understand, so I except that to fail the test as well.

**Comment by gnzlbg** [comment]
This now means that foreign exceptions are not caught by the `catch_unwind` at the thread boundary right ? I'm not sure if letting all exceptions unwind up to the thread boundary is the right thing to do, we might want to actually catch foreign exceptions there and `abort` the process explaining what happened, but I'm not sure how to handle the foreign exceptions used by `pthread_cancel` in that case (those should not be caught).

**Comment by Amanieu** [comment]
Allowing an exception to unwind to the thread boundary will cause an abort anyways at the point of the throw since it won't be able to find a catch handler.

File: src/test/run-make-fulldeps/foreign-exceptions/foo.cpp
Comment (dgrunwald) [inline src/test/run-make-fulldeps/foreign-exceptions/foo.cpp]: `x.ok` is uninitialized.
This means it's undefined behavior to call the destructor without the `x.ok = ok;` assignment; which means the C++ compiler could just optimize out `x.ok` and always directly use `ok` in its place -- which might cause the test to succeed even if the `catch (...)` doesn't catch the foreign exception.

**Comment by gnzlbg** [comment]
> Allowing an exception to unwind to the thread boundary will cause an abort anyways at the point of the throw since it won't be able to find a catch handler.

I think we should have a test for the error message here. Also, this is not always the case, e.g., when the exception thrown performs forced unwinding, e.g., the exception thrown by cancellation points when `pthread_cancel`, `abort` won't be called (which is good).

File: src/test/run-make-fulldeps/foreign-exceptions/foo.cpp
Comment (Amanieu) [reply src/test/run-make-fulldeps/foreign-exceptions/foo.cpp]: Oops, it seems I got my C++ initialization rules mixed up.

**Comment by Amanieu** [comment]
> I think we should have a test for the error message here. Also, this is not always the case, e.g., when the exception thrown performs forced unwinding, e.g., the exception thrown by cancellation points when `pthread_cancel`, `abort` won't be called (which is good).

Forced unwinds work differently from normal unwinds: there is a callback that is called on every frame which allows the caller of `_Unwind_ForcedUnwind` to choose when to stop unwinding the stack. This is what `pthread_cancel` uses to unwind to the root of the thread.

**Comment by Centril** [comment]
(It's unclear to me which team(s) owns what here, so I've nominated for all the relevant ones. Also, this interacts with the unwinding WG probably. cc @nikomatsakis @BatmanAoD)

**Comment by gnzlbg** [comment]
Since the only code that can observe this change is code that has UB I think this is strictly a T-compiler issue.

**Comment by gnzlbg** [comment]
> Forced unwinds work differently from normal unwinds: there is a callback that is called on every frame which allows the caller of _Unwind_ForcedUnwind to choose when to stop unwinding the stack. This is what pthread_cancel uses to unwind to the root of the thread.

@Amanieu in my tests, the current implementation of `catch_unwind` (not the one in this PR) catches all forced unwinds though. The one used by `pthreads` has a callback at the thread boundary that catches it inside the pthread implementation, but I was wondering what to do with other forced unwinds that do reach the thread boundary (is there anything we can do about those or a way to tell them apart from the pthread one?)

**Comment by nikomatsakis** [comment]
I definitely think this falls squarely under the nascent "wg-ffi-unwind".

**Comment by BatmanAoD** [comment]
I don't have a Windows machine available for testing; can anyone confirm whether this approach works for SEH? (Also, does anyone know what the current behavior of `catch_unwind` is on Windows when `longjmp` crosses it?)

**Comment by Amanieu** [comment]
`longjmp` on Windows is implemented using SEH unwinding, so it should behave the same way as a C++ exception.

I'm going to let bors run the tests on the Windows targets since they are all tier 1.

@bors try

**Comment by BatmanAoD** [comment]
@Amanieu In fact, @alexcrichton [fixed some logic](https://github.com/rust-lang/rust/pull/48572) along those lines some time back. I was concerned about your statement above that you expect this patch to fail on Windows, but I look forward to bors proving you wrong 😄

**Comment by retep998** [comment]
I know previously when we didn't abort on unwinding across the FFI barrier, C++ exceptions would call Rust drop code just fine and vice versa on `pc-windows-msvc`.

**Comment by Amanieu** [comment]
Does anyone know why win64-gnu doesn't just use libunwind's SEH support and rolls its own thing instead? This [code](https://github.com/gcc-mirror/gcc/blob/master/libgcc/unwind-seh.c#L251) in libgcc's SEH unwind implementation causes C++ destructors to not get called when unwinding from a Rust panic.

If there are no objections, I would like to make win64-gnu use libunwind for unwinding instead of raw SEH. This is what gcc & clang currently do for that target.

cc @vadimcn @alexcrichton

**Comment by vadimcn** [comment]
> Does anyone know why win64-gnu doesn't just use libunwind's SEH support and rolls its own thing instead? 

The main goal was to squeeze out dependencies on libgcc from Rust codebase (LLVM's libunwind did not exist back then).  Otherwise, Rust apps on Windows had to drag libgcc dll along with them, even if everything else was linked statically.   This is less important on Linux, because GCC libs are pretty much a part of the OS there. 

I also vaguely remember something about earlier implementation (which did use GCC's unwinder) having to jump through extra hoops to make Rust panics work, because normally they'd be considered "foreign exceptions" by the GCC personality routine, so it wouldn't allow to catch them (which we needed to implement `catch_unwind()`).
And the implementation before *that* was based on straight C++ exceptions (by calling out to a C++ library).

> If there are no objections, I would like to make win64-gnu use libunwind for unwinding instead of raw SEH.

If it can be done in a clean way, I don't see why not.

**Comment by mati865** [comment]
>  Otherwise, Rust apps on Windows had to drag libgcc dll along with them

At least for MinGW libunwind will link to `gcc` and `gcc_s` (this one is the shared library) by default since it gets `_Unwind` symbols from them.
One can avoid it by linking to compiler-rt instead but I don't know how well it works with compiler other than Clang.

> If there are no objections, I would like to make win64-gnu use libunwind for unwinding instead of raw SEH

Libunwind definetly supports this but it has not tested as much as Linux implementations.
If it happens to work just fine when mixing it with objects compiled by standard mingw gcc  that change would be beneficial IMO.

**Comment by vadimcn** [comment]
Another thing to consider though: 
Does Rust Project have the ambition to ever be able to compile its own toolchain without involving a C++ compiler? (Or, at least, not depending on GCC specifically)

If the answer is "yes", we should be moving more code towards pure-Rust implementation, the opposite of this change.

That's more of a philosophical/policy decision, so I'll let the core team make that call.

**Comment by Amanieu** [comment]
I would argue that the libunwind API is language agnostic and not tied to C++. It's also not strictly tied to GCC since LLVM also has its own implementation.

On the other hand the windows-msvc targets directly rely on the CRT for exception handling (in particular for the `_CxxFrameHandler3` personality function).

**Comment by vadimcn** [comment]
> I would argue that the libunwind API is language agnostic and not tied to C++. It's also not strictly tied to GCC since LLVM also has its own implementation.

Sure, but then perhaps we should be linking to that, instead of the libgcc implementation.   
And what about calling out to `_GCC_specific_handler`?   I don't see how that could be called "language-agnostic".

I would also argue that at some point we'll want to re-implement libunwind in Rust too.  There are already Rust re-implementations of compiler-rt and libm, so at this point libunwind is pretty much the only piece of the Rust runtime that requires a C++ compiler.  (The biggest elephant in the room is, of course, LLVM itself, but then there's Cranelift...)

> On the other hand the windows-msvc targets directly rely on the CRT for exception handling (in particular for the _CxxFrameHandler3 personality function).

Yeah, I was not a big fan on that implementation either.   AFAIR, it turned out that LLVM had some hard-coded assumptions based on the name of personality function.   Fixing this properly would require significant changes on LLVM side, so, in the interest of expediency, this code went in.
Still, I don't think this is such a great argument for regressing in other areas.

**Comment by alexcrichton** [comment]
I don't think I'd have much to add over what @vadimcn already mentioned. In general we want Rust (even for MinGW) to be easy to build and not have weird binary dependencies when we distribute it. So long as that's all taken care of I don't think it really matters how unwinding works, especially as we tweak it to be more consistent across platforms.

**Comment by Amanieu** [comment]
So x86_64-pc-windows-gnu now works correctly. There is no change in dependencies since we were depending on libgcc anyways.

I noticed that the `eh_unwind_resume` lang item is no longer used anywhere. Should I remove it?

There's still a few remaining issues that I am looking into:
- i686-pc-windows-gnu seems to be crashing when throwing a C++ exception.
- *-pc-windows-msvc seem to be skipping a C++ destructor when unwinding, which is strange...

**Comment by ollie27** [comment]
>     * *-pc-windows-msvc seem to be skipping a C++ destructor when unwinding, which is strange...

You'll need to pass [`/EHs`](https://docs.microsoft.com/en-us/cpp/build/reference/eh-exception-handling-model?view=vs-2019) to `cl.exe` when building `foo.cpp`.

**Comment by Amanieu** [comment]
> You'll need to pass [`/EHs`](https://docs.microsoft.com/en-us/cpp/build/reference/eh-exception-handling-model?view=vs-2019) to `cl.exe` when building `foo.cpp`.

That seems to have worked, thanks!

So the current status is:
- All libunwind platforms now work fine.
- MSVC was already working fine.
- win64-gnu now works fine.
- win32-gnu is tricky because we must have only one libunwind/libgcc linked in. This causes issues if you have both a static libgcc and a dynamic libgcc brought in by libstdc++. Hopefully my latest commit should handle that.
- The code in `emcc.rs` looks wrong, since we only want `catch_unwind` to catch Rust panics, not all exceptions. However I don't know how to setup this platform for testing. @alexcrichton could you provide some advice here?

@bors try

**Comment by Amanieu** [comment]
I'm dropping the change which removes `eh_unwind_resume` since it's more trouble than it's worth.

@bors try

**Comment by Amanieu** [comment]
Alright, all the Windows targets work now. I think this is pretty much ready to merge.

The only outstanding issue is proper handling of unwinding with emscripten, but I'm not really sure how to solve that.

**Comment by Amanieu** [comment]
I'm going to defer any work on emscripten until #65832 is merged.

File: src/libpanic_unwind/gcc.rs
Comment (Amanieu) [reply src/libpanic_unwind/gcc.rs]: AFAIK libgcc only supports SEH unwinding for x86_64. Does MinGW even have targets for ARM Windows?

Comment (mati865) [reply src/libpanic_unwind/gcc.rs]: Mingw-w64 (often called MinGW) does support ARM and Aarch64 but there are no official prebuilt toolchains available.
Unofficial toolchains like [llvm-mingw](https://github.com/mstorsjo/llvm-mingw/blob/7edc154d50ac50f84c5309487b31ba43086f2889/wrappers/clang-target-wrapper.sh#L39-L49) support more targets.

File: src/test/run-make-fulldeps/foreign-exceptions/foo.cpp
Comment (ollie27) [inline src/test/run-make-fulldeps/foreign-exceptions/foo.cpp]: Shouldn't this be checked before the `return;`?

**Comment by ollie27** [comment]
>     * MSVC was already working fine.

I should point out that MSVC currently uses C++ exceptions with type `unsigned __int64*` for Rust unwinding. If you specify the correct type then C++ can catch Rust unwinds with `catch (unsigned __int64* e)` (or `catch (void* e)`). Similarly Rust can catch C++ exceptions with `catch_unwind` if they have type `unsigned __int64*`. So I wouldn't say that it's working completely correctly yet.

File: src/test/run-make-fulldeps/foreign-exceptions/foo.cpp
Comment (Amanieu) [reply src/test/run-make-fulldeps/foreign-exceptions/foo.cpp]: Fixed, thanks.

**Comment by Amanieu** [comment]
> I should point out that MSVC currently uses C++ exceptions with type `unsigned __int64*` for Rust unwinding. If you specify the correct type then C++ can catch Rust unwinds with `catch (unsigned __int64* e)` (or `catch (void* e)`). Similarly Rust can catch C++ exceptions with `catch_unwind` if they have type `unsigned __int64*`. So I wouldn't say that it's working completely correctly yet.

I just changed the SEH implementation to use a `struct rust_panic` instead, except that I encoded the type name without name mangling so that it is impossible to catch from C++.

**Comment by Amanieu** [comment]
All done!

Emscripten support should work in theory. It basically just uses the C++ exception API for unwinding instead of calling libunwind directly so I was able to test it on Linux with no issues.

**Comment by Amanieu** [comment]
I updated the PR description with details of the new behavior of foreign exceptions unwinding through Rust code, and Rust panics unwinding through FFI.

File: src/libpanic_unwind/gcc.rs
Comment (alexcrichton) [inline src/libpanic_unwind/gcc.rs]: Would it be possible perhaps while you're here to use `cfg_if!` to clean up this `#[cfg]` clause?

File: src/libpanic_unwind/emcc.rs
Comment (alexcrichton) [inline src/libpanic_unwind/emcc.rs]: FWIW I don't think this is tested on CI right now since emscripten is hardwired to `panic=abort` right now. This is being updated in https://github.com/rust-lang/rust/pull/65832, but just wanted to note this in case you were assuming CI would test the validity of the changes to this file

File: src/libpanic_unwind/seh.rs
Comment (alexcrichton) [inline src/libpanic_unwind/seh.rs]: I'm actually sort of surprised that the `bootstrap` stuff is needed here, since this crate both defines catches and throws of panics.

Did this come up during testing though and it didn't pass without the flags? If it works it works and that's fine, just curious!

Comment (alexcrichton) [reply src/libpanic_unwind/seh.rs]: Oh this may also be connected to lang item switcharoos which definitely would necessitate `#[cfg(bootstrap)]`

File: src/test/run-make-fulldeps/foreign-exceptions/foo.rs
Comment (alexcrichton) [inline src/test/run-make-fulldeps/foreign-exceptions/foo.rs]: This `unreachable` is inside of a `catch_unwind`, so that could cause some weirdness perhaps?

Would it be better to instead of using `unreachable!` to use something like a locally defined `abort(msg: &str)` function which calls `std::process::exit`?

Comment (alexcrichton) [reply src/test/run-make-fulldeps/foreign-exceptions/foo.rs]: To clarify, I think this test has enough checks that if this `unreachable!` is hit the test will still fail, but it  might help improving things to abort instead of relying on panics for test failure

File: src/test/run-make-fulldeps/foreign-exceptions/foo.rs
Comment (alexcrichton) [inline src/test/run-make-fulldeps/foreign-exceptions/foo.rs]: Should this function be tagged with `#[unwind(allowed)]` (for future proofing?)

File: src/test/run-make-fulldeps/foreign-exceptions/foo.rs
Comment (alexcrichton) [inline src/test/run-make-fulldeps/foreign-exceptions/foo.rs]: Should this be tagged with `#[unwind(allowed)]`? (same as below)

**Review by alexcrichton — COMMENTED** [review]
Thanks for the changes here @Amanieu, this is looking great! 

The [OP description](https://github.com/rust-lang/rust/pull/65646#issue-330187431) was also very descriptive, thanks! Can you also explain in a bit more detail what's going on with MinGW? All the changes here I was able to match up with the description but I want to just double check my interpretation of MinGW.

Also you mention `longjmp` on MSVC, where I believe we intentionally don't want to run Rust destructors on `longjmp` (it doesn't anywhere else so I figured it shouldn't on MSVC). I didn't really see any changes which would lead to that though, so I'm curious is this something that's intented as a result of this PR?

File: src/test/run-make-fulldeps/foreign-exceptions/foo.rs
Comment (Amanieu) [reply src/test/run-make-fulldeps/foreign-exceptions/foo.rs]: I think it's fine to leave it as it is: the code below verifies that the panic object is a `Box<i32>` and the unreachable message will appear in the stderr output of the test.

File: src/libpanic_unwind/seh.rs
Comment (Amanieu) [reply src/libpanic_unwind/seh.rs]: Yes, it's because I've changed the exception type from `unsigned __int64*` to `struct rust_panic` (i.e. I've removed one layer of indirection). This matches up with the changes in `codegen_msvc_try`.

File: src/libpanic_unwind/gcc.rs
Comment (Amanieu) [reply src/libpanic_unwind/gcc.rs]: > The [OP description](https://github.com/rust-lang/rust/pull/65646#issue-330187431) was also very descriptive, thanks! Can you also explain in a bit more detail what's going on with MinGW? All the changes here I was able to match up with the description but I want to just double check my interpretation of MinGW.

For 32-bit MinGW there is no change (apart from the handling of foreign exceptions in the personality function). For 64-bit MinGW we want to use the same exception mechanism as C++ code (libunwind) instead of rolling our own thing. Conveniently, libunwind already has a translation layer (`_GCC_specific_handler`) which wraps a libunwind-style personality function and hides all the ugly SEH details. This enables:
- us to run Rust destructors when a C++ exception unwinds, but not for arbitrary SEH exceptions (such as the one used by `longjmp`).
- C++ destructors to run when a Rust panic unwinds, since `_GCC_specific_handler` only passes exceptions with code `STATUS_GCC_THROW` and `STATUS_GCC_UNWIND` to the libunwind-style personality function.

> Also you mention `longjmp` on MSVC, where I believe we intentionally don't want to run Rust destructors on `longjmp` (it doesn't anywhere else so I figured it shouldn't on MSVC). I didn't really see any changes which would lead to that though, so I'm curious is this something that's intented as a result of this PR?

Actually `longjmp` does run destructors (both C++ and Rust) because it uses a SEH exception to unwind. This was already the case before this PR. However `_CxxFrameHandler3` specifically ignores `longjmp` exceptions for `catch` blocks.

This behavior is documented in MSVC, and is allowed by the C++ standard since it is UB to unwind across frames with destructors.

**Comment by Amanieu** [comment]
> The [OP description](https://github.com/rust-lang/rust/pull/65646#issue-330187431) was also very descriptive, thanks! Can you also explain in a bit more detail what's going on with MinGW? All the changes here I was able to match up with the description but I want to just double check my interpretation of MinGW.

For 32-bit MinGW there is no change (apart from the handling of foreign exceptions in the personality function). For 64-bit MinGW we want to use the same exception mechanism as C++ code (libunwind) instead of rolling our own thing. Conveniently, libunwind already has a translation layer (`_GCC_specific_handler`) which wraps a libunwind-style personality function and hides all the ugly SEH details. This enables:
- us to run Rust destructors when a C++ exception unwinds, but not for arbitrary SEH exceptions (such as the one used by `longjmp`).
- C++ destructors to run when a Rust panic unwinds, since `_GCC_specific_handler` only passes exceptions with code `STATUS_GCC_THROW` and `STATUS_GCC_UNWIND` to the libunwind-style personality function.
I've updated the code with your suggestions.

> Also you mention `longjmp` on MSVC, where I believe we intentionally don't want to run Rust destructors on `longjmp` (it doesn't anywhere else so I figured it shouldn't on MSVC). I didn't really see any changes which would lead to that though, so I'm curious is this something that's intented as a result of this PR?

Actually `longjmp` does run destructors (both C++ and Rust) because it uses a SEH exception to unwind. This was already the case before this PR. However `_CxxFrameHandler3` specifically ignores `longjmp` exceptions for `catch` blocks.

This behavior is documented in MSVC, and is allowed by the C++ standard since it is UB to unwind across frames with destructors.

**Comment by alexcrichton** [comment]
This seems reasonable enough to me, I'm pretty uncomfortable about `longjmp` and MSVC but it's not a reason to block this.

r? @nikomatsakis 

(sorry I don't know who best to transfer review to)

This is basically an r+ from me on the technical bits, but on the bits about whether we want to do this at all for unwinding and such (basically wg-ffi-unwind), I'm deferring the actual decision to merge to them. Sorry Niko but I know you're at least somewhat related, feel free to transfer r? to someone else in that group!

**Comment by nikomatsakis** [comment]
@alexcrichton 

I can do final review, thanks.

> This seems reasonable enough to me, I'm pretty uncomfortable about longjmp and MSVC but it's not a reason to block this.

Specifically, you're uncomfortable about having longjmp *run* destructors on MSVC?

**Comment by nikomatsakis** [comment]
@Amanieu 

> This behavior is documented in MSVC, and is allowed by the C++ standard since it is UB to unwind across frames with destructors.

What do you mean by "unwind" here -- I guess you specifically mean it is UB to longjmp across a frame with destructors? (Or, more generally, to unwind via any mechanism *other* than in response to a C++ throw?)

**Comment by Amanieu** [comment]
> What do you mean by "unwind" here -- I guess you specifically mean it is UB to longjmp across a frame with destructors? (Or, more generally, to unwind via any mechanism _other_ than in response to a C++ throw?)

Sorry yes, I meant that it is UB longjmp across a frame with destructors. If we apply this rule to Rust as well, then we can simply ignore the platform differences in longjmp running destructors since it is UB anyways.

**Comment by alexcrichton** [comment]
@nikomatsakis yes that's the part I'm uncomfortable about. If we already do it though and we don't try to document that it works, then it seems like maintaining the status quo which is fine

**Comment by nikomatsakis** [comment]
OK, @Amanieu, can you confirm: even before this PR lands, longjmp on windows already runs destructors on Rust frames that it jumps over? (unwinds through)

(That's what I took your comment to mean)

**Comment by Amanieu** [comment]
> OK, @Amanieu, can you confirm: even before this PR lands, longjmp on windows already runs destructors on Rust frames that it jumps over? (unwinds through)

Yes, but only on windows-msvc, windows-gnu doesn't run destructors when unwinding with a longjmp. In both cases the longjmp behavior is the same as before this PR.

**Review by nikomatsakis — APPROVED** [review]
I have a few minor suggestions.

File: src/libpanic_unwind/emcc.rs
Comment (Amanieu) [reply src/libpanic_unwind/emcc.rs]: It's probably not even necessary here, I just copied it from `seh.rs`.

File: src/libpanic_unwind/dwarf/eh.rs
Comment (bjorn3) [reply src/libpanic_unwind/dwarf/eh.rs]: I believe C++ does catch foreign exceptions using `try { ... } catch(exception e) { ... }`.

Comment (Amanieu) [reply src/libpanic_unwind/dwarf/eh.rs]: Yes but `catch_unwind` needs to return a `Box<Any>` when it catches an exception, which we can only do for Rust panics since those hold the `Box<Any>` as the exception payload.

Comment (bjorn3) [reply src/libpanic_unwind/dwarf/eh.rs]: For other exceptions I think we could use some opaque `ForeignExceptionPayload` type containing the payload received from the unwinder as type behind the `Box<Any>`. And maybe even unwrap it again when calling `std::panic::resume_unwind`.

Comment (BatmanAoD) [reply src/libpanic_unwind/dwarf/eh.rs]: @bjorn3 Not that this impacts your overall point, but C++ should only be able to catch foreign exceptions with `catch {` or `catch (null*) {`

Comment (Amanieu) [reply src/libpanic_unwind/dwarf/eh.rs]: (small correction regarding the C++ syntax: it's `catch (...)` that can catch foreign exceptions)

That's unfortunately not possible for several reasons:
- The foreign exception object is completely opaque (since we don't know its real type) and the only operations we can do with it is rethrow it or free it.
- The foreign exception object has a lifetime limited to the `catch` clause that caught it. In Rust, the catch block is entirely hidden in the `intrinsics::try` compiler builtin.

In summary, we only have 2 choices: let foreign exceptions silently unwind through `catch_unwind`, or catch the foreign exception and immediately free it, then report that "some foreign exception happened" through a `Box<ForeignException>`.`ForeignException` would be a ZST since at that point it is too late to rethrow the exception (we've left the `catch` block).

Comment (nikomatsakis) [reply src/libpanic_unwind/dwarf/eh.rs]: Or abort.

Comment (nikomatsakis) [reply src/libpanic_unwind/dwarf/eh.rs]: I don't think this PR is the right place to resolve this question though

**Review by nikomatsakis — APPROVED** [review]

**Comment by nikomatsakis** [comment]
@bors r+

Thanks @Amanieu, this is great

File: src/libpanic_unwind/seh.rs
Comment (nikomatsakis) [inline src/libpanic_unwind/seh.rs]:
```suggestion
#[cfg_attr(bootstrap, msvc_try_filter)]
```

File: src/libpanic_unwind/dwarf/eh.rs
Comment (gnzlbg) [reply src/libpanic_unwind/dwarf/eh.rs]: C++ can catch foreign exceptions with both `catch(...) { ... }` _and_ `catch(__foreign_exception& e) { ... }` as well, where the `__foreign_exception` types can be exposed in `<exception>` header and can have sub-classes like `__java_exception` - which types are available is implementation-defined.

See, for example, [the Itanium C++ EH ABI section 1.6.4](https://itanium-cxx-abi.github.io/cxx-abi/abi-eh.html#base-personality):  

> Example: Foreign Exceptions in C++. In C++, foreign exceptions can be caught by a `catch(...)` statement. They can also be caught as if they were of a `__foreign_exception` class, defined in `<exception>`. The `__foreign_exception` may have subclasses, such as `__java_exception` and `__ada_exception`, if the runtime is capable of identifying some of the foreign languages.


>> let foreign exceptions silently unwind through catch_unwind, or catch the foreign exception and immediately free it, then report that "some foreign exception happened" through a
>
> Or abort.
>
> I don't think this PR is the right place to resolve this question though

Shall we open an issue in project-ffi-unwind to track this ? Or what would be the appropriate forum to discuss it ? I think we should try to at least properly document this behavior somewhere at some point, and that it would be worth to continue discussing which options we have for defining it somewhere.

File: src/libpanic_unwind/seh.rs
Comment (bjorn3) [reply src/libpanic_unwind/seh.rs]:
```suggestion
#[cfg_attr(bootstrap, lang = "msvc_try_filter")]
```

To fix https://github.com/rust-lang/rust/pull/65646#issuecomment-549032221

Comment (Amanieu) [reply src/libpanic_unwind/seh.rs]: Thanks. Note to self: allow time to wake up properly before fixing code.

**Pull request merged**
